qrenderdoc/Windows/Dialogs/PerformanceCounterSelection.cpp

```diff
@@ -0,0 +1,290 @@
+
+const char *ToString(CounterFamily family)
+{
+  switch(family)
+  {
+    case CounterFamily::AMD: return "AMD";
+    case CounterFamily::Generic: return "Generic";
+    case CounterFamily::Intel: return "Intel";
+    case CounterFamily::NVIDIA: return "NVIDIA";
+    case CounterFamily::Unknown: return "Unknown";
+  }
+
+  return nullptr;
+}
+}
+
+PerformanceCounterSelection::PerformanceCounterSelection(ICaptureContext &ctx, QWidget *parent)
+    : QDialog(parent), ui(new Ui::PerformanceCounterSelection), m_Ctx(ctx)
+{
+  ui->setupUi(this);
+
+  connect(ui->counterTree, &QTreeWidget::itemEntered, [&](QTreeWidgetItem *item, int) -> void {
+    const auto d = item->data(0, Qt::UserRole + 1);
+
+    if(d.isValid())
+    {
+      ui->counterDescription->setText(
+          QString(QLatin1String("<b>%1</b><hr>%2")).arg(item->text(0)).arg(d.toString()));
+    }
+  });
+
+  connect(ui->save, &QPushButton::pressed, [&]() -> void {
+    QString filename = RDDialog::getSaveFileName(this, tr("Save File"), QDir::homePath(),
+                                                 tr("Performance Counter Settings (*.json)"));
+
+    if(filename.isEmpty())
+      return;
+
+    QVariantList counterIds;
+    for(auto v : m_SelectedCounters.keys())
+    {
+      const Uuid uuid = m_CounterToUuid[v];
+      QVariantList e;
+
+      for(const auto b : uuid.bytes)
+      {
+        e.append(b);
+      }
+
+      counterIds.append(QVariant{e});
+    }
+
+    QVariantMap doc;
+    doc[QLatin1String{"counters"}] = counterIds;
+
+    QFile f(filename);
+    if(f.open(QIODevice::WriteOnly | QIODevice::Truncate | QIODevice::Text))
+    {
+      SaveToJSON(doc, f, JSON_ID, JSON_VER);
+    }
+    else
+    {
+      RDDialog::critical(this, tr("Error saving config"),
+                         tr("Couldn't open path %1 for write.").arg(filename));
+    }
+  });
+
+  connect(ui->load, &QPushButton::pressed, [&]() -> void {
+    QString filename = RDDialog::getOpenFileName(this, tr("Load file"), QDir::homePath(),
+                                                 tr("Performance Counter Settings (*.json)"));
+
+    if(filename.isEmpty())
+      return;
+
+    QVariantMap doc;
+    QFile f(filename);
+    if(f.open(QIODevice::ReadOnly | QIODevice::Text))
+    {
+      LoadFromJSON(doc, f, JSON_ID, JSON_VER);
+
+      std::set<Uuid> selectedCounters;
+
+      QVariantList counters = doc[QLatin1String{"counters"}].toList();
+
+      for(const auto &counter : counters)
+      {
+        QVariantList bytes = counter.toList();
+        Uuid uuid;
+
+        /// TODO assert counter.size () == 4
+
+        for(int i = 0; i < 4; ++i)
+        {
+          uuid.bytes[i] = bytes[i].toUInt();
+        }
+
+        selectedCounters.insert(uuid);
+      }
+
+      // We we walk over the complete tree, and toggle everything so it
+      // matches the settings
+      QTreeWidgetItemIterator it(ui->counterTree);
+      while(*it)
+      {
+        const auto id = (*it)->data(0, Qt::UserRole + 2);
```

Comment:
Since these roles are used in a couple of places, it'd be nice to pull them out into `static const int` constants at the top of the `.cpp` file, just so it's immediately obvious that `Qt::UserRole + 2` is the counter ID.